### PR TITLE
feat: Updates CRT to 0.5.0+

### PR DESCRIPTION
**Pull request by epau**
## Issue \#
https://github.com/awslabs/aws-sdk-swift/issues/756

## Description of changes
Updates CRT dependency to 0.5.0 and fixes smithy-swift to make it compatible. 

## Scope

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Packages/ClientRuntime/Sources/Config/DefaultSDKRuntimeConfiguration.swift
Comment (epau) [inline Packages/ClientRuntime/Sources/Config/DefaultSDKRuntimeConfiguration.swift:8]: We should rename this struct and the protocol to something more descriptive and without "SDK" in the name.
I'll file an issue to address that

File: Packages/ClientRuntime/Sources/Networking/Http/CRT/SDKDefaultIO.swift
Comment (epau) [inline Packages/ClientRuntime/Sources/Networking/Http/CRT/SDKDefaultIO.swift:16]: We should rename this type to something more descriptive and without "SDK" in the name. 
I'll file an issue to address this.

File: Packages/ClientRuntime/Sources/Networking/Http/SdkHttpRequest.swift
Comment (epau) [inline Packages/ClientRuntime/Sources/Networking/Http/SdkHttpRequest.swift:42]: This `bufferSize` param was never used, so its now removed.

File: Packages/ClientRuntime/Sources/Networking/Http/SdkHttpRequest.swift
Comment (epau) [inline Packages/ClientRuntime/Sources/Networking/Http/SdkHttpRequest.swift:78]: `path` is no longer optional, so the nil coalescing is no longer needed

File: Packages/ClientRuntime/Sources/Networking/Streaming/ByteStream.swift
Comment (epau) [inline Packages/ClientRuntime/Sources/Networking/Streaming/ByteStream.swift:12]: I couldn't find a good use case to use this `reader` variant. It appears it was never code generated and `ByteBuffer` is sufficient for streaming requests. Therefore, this variant is removed and the `DataStreamReader` class is also removed.

Please do let me know if you think we should keep it though.

File: Packages/ClientRuntime/Sources/Retries/SDKRetryer.swift
Comment (epau) [inline Packages/ClientRuntime/Sources/Retries/SDKRetryer.swift:112]: This is the same behavior currently today, where we end up just passing in `nil` for this closure since we cannot convert swift closures to c closures due to constraints around c closures. 

I've talked to @waahm7 about this and filed https://github.com/awslabs/aws-crt-swift/issues/127 for CRT to expose this as a swift closure. Once that is done, we can update this here to pass in the provided swift closure.

File: Sources/ClientRuntime/Config/DefaultSDKRuntimeConfiguration.swift
Comment (epau) [inline Sources/ClientRuntime/Config/DefaultSDKRuntimeConfiguration.swift:8]: We should rename this to something more descriptive and without the terns "Default" and "SDK".
I'll file an issue to address this

File: Sources/ClientRuntime/Config/SDKRuntimeConfiguration.swift
Comment (epau) [inline Sources/ClientRuntime/Config/SDKRuntimeConfiguration.swift:23]: This is a good example of when adding default implementations to protocols does not make sense. 
These defaults are all specific to the "default" or "crt based" configuration, and so we should be setting these values there and not at the protocol level.

File: Sources/ClientRuntime/Networking/Http/CRT/CRTClientEngine.swift
Comment (epau) [inline Sources/ClientRuntime/Networking/Http/CRT/CRTClientEngine.swift:142]: There no longer is anything to close on the CRT side of things and so we just no-op here. 
We can remove the `close` function from the interface as well, and then delete it here but I left it in as it might sense to "close" other http engine implementations.

File: Sources/ClientRuntime/Networking/Streaming/ByteStream.swift
Comment (epau) [inline Sources/ClientRuntime/Networking/Streaming/ByteStream.swift:12]: The stream reader case was unused and seems unnecessary since everything can be handled via the ByteBuffer. Please let me know if you think this needs to stay in though.

File: Sources/ClientRuntime/Retries/SDKRetryer.swift
Comment (epau) [inline Sources/ClientRuntime/Retries/SDKRetryer.swift:105]: This maintains the existing behavior where we just pass in `nil` to generateRandom since it is a c closure. An issue has been filed with CRT to change the type exposed so that we can pass in a swift closure. Otherwise, we would need to expose the c closure type to consumers which doesn't make sense as it feels like an implementation detail leaking out.

File: Sources/ClientRuntime/Networking/Http/CRT/SDKDefaultIO.swift
Comment (jbelkins) [inline Sources/ClientRuntime/Networking/Http/CRT/SDKDefaultIO.swift:19]: Not the kind of thing that ought to be a "TODO" comment.

**Review by jbelkins — APPROVED** [review]
Approved since this gets us onto latest CRT, and the API changes they made do look like an improvement.

For the future, we should have an eye on how to separate CRT from smithy-swift altogether though.

File: Sources/ClientRuntime/Networking/Http/CRT/CRTClientEngine.swift
Comment (ganeshnj) [reply Sources/ClientRuntime/Networking/Http/CRT/CRTClientEngine.swift:142]: would suggest deleting this entirely. We can do before go GA but not later.

File: Sources/ClientRuntime/Retries/RetryError.swift
Comment (ganeshnj) [inline Sources/ClientRuntime/Retries/RetryError.swift:17]: isn't is `CommonRuntimeKit` now?

File: Sources/ClientRuntime/Networking/Http/CRT/CRTClientEngine.swift
Comment (epau) [reply Sources/ClientRuntime/Networking/Http/CRT/CRTClientEngine.swift:108]: Is it odd? I had thought this was a pretty popular formatting style in the community and is the default way SwiftFormat will format multiline guard statements. 

Would be nice if Swift would prescribe some "standard format" one of these days...

File: Sources/ClientRuntime/Networking/Http/HttpBody.swift
Comment (epau) [reply Sources/ClientRuntime/Networking/Http/HttpBody.swift:19]: iirc i needed public access to `toBytes` but maybe not. I'll double check and update accordingly

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/httpResponse/bindingTraits/HttpResponseTraitWithHttpPayload.kt
Comment (epau) [reply smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/httpResponse/bindingTraits/HttpResponseTraitWithHttpPayload.kt:32]: Yeah i agree, a simple computed var named `data` would be nice here.

File: Sources/ClientRuntime/Retries/RetryError.swift
Comment (epau) [reply Sources/ClientRuntime/Retries/RetryError.swift:17]: Negative, [CommonRuntimeKit](https://github.com/awslabs/aws-crt-swift/blob/f3ae10c9a33bdc9d04246141ef0eb5840b369205/Source/AwsCommonRuntimeKit/CommonRuntimeKit.swift) is a struct while `AwsCommonRuntimeKit` is the module. 

We should clean up these names though. It's odd to have something with a `Kit` suffix that isn't a module.

File: Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift
Comment (epau) [reply Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift:10]: Nope. The one place this was used, `ByteBuffer` is sufficient https://github.com/awslabs/smithy-swift/pull/497/files#diff-9ae04687d7469a19d324ef46efc65eada692f0b62c0f49890bc0a6c27afd4358L111

There doesn't seem to be any added functionality provided by the `DataStreamReader` nor `StreamReader` (compared to `ByteBuffer`) and their purposes are unclear. We can simply rely on `ByteBuffer` for streaming data.

File: Sources/ClientRuntime/Networking/Streaming/StreamReader.swift
Comment (epau) [reply Sources/ClientRuntime/Networking/Streaming/StreamReader.swift:6]: See this comment https://github.com/awslabs/smithy-swift/pull/497#discussion_r1055682148

File: Sources/ClientRuntime/Networking/Http/HttpBody.swift
Comment (epau) [reply Sources/ClientRuntime/Networking/Http/HttpBody.swift:19]: Confirmed we need to access `toBytes` outside of this module. 

We no longer need to the `toAwsInputStream` because CRT changed how input streams are defined. Now its a protocol, `IStreamable` that `ByteBuffer` conforms to.

File: Sources/ClientRuntime/Networking/Streaming/ByteStream.swift
Comment (ganeshnj) [reply Sources/ClientRuntime/Networking/Streaming/ByteStream.swift:12]: I would push back on removing this as we don't the side effect yet. Let's keep this.

This can be cleanedup when doing streaming improvements.

**Review by ganeshnj — CHANGES_REQUESTED** [review]

File: Sources/ClientRuntime/Retries/SDKRetryer.swift
Comment (waahm7) [inline Sources/ClientRuntime/Retries/SDKRetryer.swift:28]: I have already updated this.

File: Sources/ClientRuntime/Networking/Http/CRT/CRTClientEngine.swift
Comment (ganeshnj) [reply Sources/ClientRuntime/Networking/Http/CRT/CRTClientEngine.swift:108]: Interesting, looks new to me.

File: Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift
Comment (ganeshnj) [inline Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift:52]:
```suggestion
            let ptr = UnsafeMutablePointer<UInt8>.allocate(capacity: count)
            let buffer = UnsafeMutableBufferPointer(start: ptr, count: count)
            defer {
                  buffer.deallocate()
                  data.deallocate()
            }
            let bytesRead = byteBuffer.read(buffer: buffer)
            data = Data(buffer)
```

File: Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift
Comment (ganeshnj) [inline Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift:55]: shouldn't it fail if truncation happens?

This could lead to loss of data.

File: Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift
Comment (ganeshnj) [inline Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift:83]: same if truncation happens, we are giving wrong contentLength.

**Review by ganeshnj — CHANGES_REQUESTED** [review]

File: Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift
Comment (epau) [reply Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift:55]: Yeah good call out, I'll remove the `truncatingIfNeeded`. 
When we tackle streaming improvements, we may want to update the interface to use `UIInt64` as well

File: Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift
Comment (epau) [reply Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift:52]: hmm can you explain this change? from my understand, it is only required that we `deallocate` the `UnsafeMutablePointer` (`ptr`).

File: Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift
Comment (waahm7) [inline Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift:52]: We can replace all of this with data and a safer version which doesn't involve allocation/deallocation. Something like the following
```suggestion
            let data = Data(count: count)
            bufferData.withUnsafeMutableBytes { buffer in 
                 byteBuffer.read(buffer: buffer)
            }
```

Comment (waahm7) [reply Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift:52]: In general, the allocation & deallocation should only happen if we want to manually control lifetime of something. In most cases, we can avoid that.

**Review by waahm7 — CHANGES_REQUESTED** [review]

File: Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift
Comment (epau) [reply Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift:52]: 🤯 
awesome! will update to this

**Review by waahm7 — APPROVED** [review]

File: Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift
Comment (ganeshnj) [reply Sources/ClientRuntime/Networking/Streaming/DataStreamReader.swift]: my bad, didn't realize it was BufferPointer, makes sense.

**Review by ganeshnj — APPROVED** [review]

**Comment by epau** [comment]
Waiting until after next release to merge, since this includes a big update from CRT

**Pull request merged**
